vllm_v1/tokenizer/detokenizer.py

```diff
@@ -0,0 +1,158 @@
+        self.encoder = msgspec.msgpack.Encoder()
+        self.decoder = msgspec.msgpack.Decoder(RequestData)
+
+        self.tokenizer = get_tokenizer(tokenizer_name)
+        self.requests: Dict[str, RequestState] = {}
+
+    def run(self):
+        self.context = zmq.Context()
+        self.pull_socket = self.context.socket(zmq.PULL)
+        self.pull_socket.bind(f"tcp://*:{self.port1}")
+        self.push_socket = self.context.socket(zmq.PUSH)
+        self.push_socket.bind(f"tcp://*:{self.port2}")
+
+        while True:
+            message = self.pull_socket.recv()
+            if message == b"":
+                # Terminate signal.
+                break
+            data = self.decoder.decode(message)
+
+            for req_id in data.free_request_ids:
+                self.free(req_id)
+
+            req_ids: List[str] = []
+            detokenized_texts: List[str] = []
+            num_reqs = len(data.request_ids)
+            for i in range(num_reqs):
+                req_id = data.request_ids[i]
+                req_ids.append(req_id)
+                if req_id not in self.requests:
+                    self.add_request(
+                        request_id=req_id,
+                        prompt_token_ids=data.prompt_token_ids[i],
+                        skip_special_tokens=data.skip_special_tokens[i],
+                        spaces_between_special_tokens=data.
+                        spaces_between_special_tokens[i],
+                    )
+                new_str = self.detokenize(req_id, data.new_token_ids[i])
+                detokenized_texts.append(new_str)
+
+            detokenized = DetokenizedData(
+                request_ids=req_ids,
+                detokenized_texts=detokenized_texts,
+            )
+            self.push_socket.send(self.encoder.encode(detokenized),
+                                  flags=zmq.NOBLOCK)
+
+    def add_request(
+        self,
+        request_id: str,
+        prompt_token_ids: List[int],
+        skip_special_tokens: bool,
+        spaces_between_special_tokens: bool,
+    ) -> None:
+        tokens, prefix_offset, read_offset = convert_prompt_ids_to_tokens(
+            tokenizer=self.tokenizer,
+            prompt_ids=prompt_token_ids,
+            skip_special_tokens=skip_special_tokens,
+        )
+        self.requests[request_id] = RequestState(
+            req_id=request_id,
+            token_ids=prompt_token_ids,
+            tokens=tokens,
+            prefix_offset=prefix_offset,
+            read_offset=read_offset,
+            skip_special_tokens=skip_special_tokens,
+            spaces_between_special_tokens=spaces_between_special_tokens,
+        )
+
+    def free(self, request_id: str) -> None:
+        del self.requests[request_id]
+
+    def detokenize(self, request_id: str, new_token_ids: List[int]) -> str:
+        # TODO(woosuk): This method becomes very inefficient when the number of
+        # new_token_ids is more than 1. We need to optimize this.
```

Comment:
Is this already very inefficient in v0 or this is new in v1?

Reply:
To my understanding, the incremental de-tokenization is efficient when we have to de-tokenize every step, while the regular de-tokenization is more efficient when handling longer output lengths.

In v0, the incremental de-tokenizer is preferable, since we de-tokenize the tokens every step. However, in this v1 prototype, we do NOT de-tokenize every step, so this incremental de-tokenizer can be inefficient.